self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb

```diff
@@ -1567,158 +1717,384 @@
+     "output_type": "stream",
+     "text": [
+      "Overwriting online-retail-clv-3M/Dockerfile\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%writefile {MODEL_NAME}/Dockerfile\n",
+    "# Specifies base image and tag.\n",
+    "# https://cloud.google.com/vertex-ai/docs/training/pre-built-containers\n",
+    "# us-docker.pkg.dev/vertex-ai/training/tf-cpu.2-3:latest\n",
+    "FROM gcr.io/deeplearning-platform-release/tf2-cpu.2-3\n",
     "\n",
-    "model.fit(training_ds,\n",
-    "          validation_data=eval_ds,\n",
-    "          steps_per_epoch=STEPS_PER_EPOCH,\n",
-    "          epochs=N_CHECKPOINTS,\n",
-    "          callbacks=[[tensorboard_callback,\n",
-    "                      earlystopping_callback,\n",
-    "                      checkpoint_callback]])\n",
+    "# Sets the container working directory.\n",
+    "WORKDIR /root\n",
     "\n",
-    "print(model.evaluate(test_ds))\n",
+    "# Copies the requirements.txt into the container to reduce network calls.\n",
+    "COPY requirements.txt .\n",
+    "# Installs additional packages.\n",
+    "RUN pip3 install -U -r requirements.txt\n",
     "\n",
-    "tmpdir = tempfile.mkdtemp()\n",
+    "# Copies the trainer code to the docker image.\n",
+    "COPY . /trainer\n",
     "\n",
-    "# Export Keras model in TensorFlow SavedModel format.\n",
-    "model.save(tmpdir)\n",
+    "# Sets the container working directory.\n",
+    "WORKDIR /trainer\n",
     "\n",
-    "# Annotate and save TensorFlow SavedModel with Explainable metadata to GCS.\n",
-    "builder = SavedModelMetadataBuilder(tmpdir)\n",
-    "builder.save_model_with_metadata(os.environ[\"AIP_MODEL_DIR\"])"
+    "# Sets up the entry point to invoke the trainer.\n",
+    "ENTRYPOINT [\"python\", \"-m\", \"trainer.task\"]"
    ]
   },
   {
    "cell_type": "markdown",
-   "id": "cb3ac018",
+   "id": "90d9fecd",
    "metadata": {},
    "source": [
-    "### Use Cloud Build to build and submit container to Google Cloud Container Registry"
+    "### 4. Write a `requirements.txt` file to specify additional ML code dependencies"
    ]
   },
   {
    "cell_type": "markdown",
-   "id": "71d8344f",
+   "id": "279d289c",
    "metadata": {},
    "source": [
-    "Next, you will use [Cloud Build](https://cloud.google.com/build) to build and upload your custom TensorFlow model container to [Google Cloud Container Registry](https://cloud.google.com/container-registry).\n",
+    "These are additional dependencies for your model code outside the pre-built containers needed for prediction explainability and the BigQuery TensorFlow IO reader."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1421,
+   "id": "b65f5d2d",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Overwriting online-retail-clv-3M/requirements.txt\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%writefile {MODEL_NAME}/requirements.txt\n",
+    "explainable-ai-sdk==1.3.0\n",
```

Comment:
comment: `explainable-ai-sdk` is included out of the box in the pre-built containers i believe, but it still makes sense to leave this in, _if_ we anticipate breaking changes in a version upgrade

Reply:
https://cloud.google.com/vertex-ai/docs/training/pre-built-containers

I didn't see it on here actually which was surprising. The other reason for this is the tensorflow IO package (not part of core) for the BigQuery reader. Pyarrow is a dependency for that.

Reply:
Oh I am probably wrong, sorry! It might be in the Notebooks but not in the actual pre-built containers. So leave it in, if in doubt!